src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -501,17 +507,18 @@ sliceSepByMax predicate cnt (Fold fstep finitial fextract) =
     where
 
     initial = Tuple' 0 <$> finitial
+
     step (Tuple' i r) a
-        | not (predicate a) = do
-            res <- fstep r a
-            let i1 = i + 1
-                s1 = Tuple' i1 res
-            if i1 < cnt
-            then return $ Partial 0 s1
-            else do
-                b <- fextract res
-                return $ Done 0 b
+        | not (predicate a) =
+            if i < cnt
+            then do
+                res <- fstep r a
+                let i1 = i + 1
+                    s1 = Tuple' i1 res
+                return $ Partial 0 s1
+            else Done 1 <$> fextract r
```

Comment:
I took another look at it, and it looks fine. Because when we reach the Done case we are already looking at the next element and we have checked that it is not passing the predicate, so we return it back. If it is passing the predicate then we will reach the Done 0 case where we drop the separator.

The ultimate proof is a test case for this case.